src/test/groovy/com/gooddata/executeafm/afm/OverPeriodMeasureDefinitionTest.groovy

```diff
@@ -0,0 +1,169 @@
+
+        then:
+        measure.measureIdentifier == 'mId'
+        measure.getObjQualifiers()[0].uri == '/gdc/md/projectId/obj/1'
+        measure.getObjQualifiers()[1].uri == '/gdc/md/projectId/obj/2'
+        measure.dateAttributes.size() == 2
+        measure.dateAttributes[0].periodsAgo == 2
+        measure.dateAttributes[0].attribute.uri == '/gdc/md/projectId/obj/1'
+        measure.dateAttributes[1].periodsAgo == 1
+        measure.dateAttributes[1].attribute.uri == '/gdc/md/projectId/obj/2'
+        measure.isAdHoc()
+        measure.toString()
+    }
+
+    def "test serializable"() {
+        given:
+        OverPeriodMeasureDefinition measureDefinition = readObjectFromResource("/$SINGLE_ATTRIBUTES_JSON", OverPeriodMeasureDefinition)
+        OverPeriodMeasureDefinition deserialized = SerializationUtils.roundtrip(measureDefinition)
+
+        expect:
+        that deserialized, jsonEquals(measureDefinition)
+    }
+
+    def "should throw when attributes is null"() {
+        when:
+        new OverPeriodMeasureDefinition('mId', null)
+        then:
+        thrown(NullPointerException)
+    }
+
+    def "should throw when attributes are empty"() {
+        when:
+        new OverPeriodMeasureDefinition('mId', [])
+        then:
+        thrown(IllegalArgumentException)
+    }
+
+    def "should throw when measure identifier is null"() {
+        when:
+        new OverPeriodMeasureDefinition(null, [new OverPeriodDateAttribute(
+                new UriObjQualifier('/gdc/md/projectId/obj/1'), 2)])
+        then:
+        thrown(NullPointerException)
+    }
+
+    def "should throw when getting single used object qualifier"() {
+        when:
+        OverPeriodMeasureDefinition measure = new OverPeriodMeasureDefinition('mId', [
+                new OverPeriodDateAttribute(new IdentifierObjQualifier('id1'), -2),
+        ])
+        measure.getObjQualifier()
+        then:
+        thrown(UnsupportedOperationException)
+    }
+
+    def "should throw when try to copy with uri"() {
+        when:
+        OverPeriodMeasureDefinition measure = new OverPeriodMeasureDefinition('mId', [
+                new OverPeriodDateAttribute(new IdentifierObjQualifier('id1'), -2),
+        ])
+        measure.withObjUriQualifier(new UriObjQualifier("/gdc/md/projectId/obj/1"))
+        then:
+        thrown(UnsupportedOperationException)
+    }
+
+    def "should copy with uri converter"() {
+        when:
+        OverPeriodMeasureDefinition measure = new OverPeriodMeasureDefinition('mId', [
+                new OverPeriodDateAttribute(new IdentifierObjQualifier('id1'), -2),
+                new OverPeriodDateAttribute(new IdentifierObjQualifier('id2'), -3),
+                new OverPeriodDateAttribute(new UriObjQualifier('/gdc/md/projectId/obj/3'), -4)
+        ])
+        def copy = measure.withObjUriQualifiers({ identifierQualifier ->
+            def qualifiersMap = [
```

Comment:
consider to declare the mapping outside of the closure, it could imporve readability

Reply:
Fixed.